atef/tools.py

```diff
@@ -0,0 +1,200 @@
+from __future__ import annotations
+
+import asyncio
+import dataclasses
+import re
+import shutil
+import sys
+import typing
+from dataclasses import field
+from typing import Any, ClassVar, Dict, List, Type, TypedDict, TypeVar, Union
+
+from . import serialization
+from .check import Result, Severity
+
+#: Arguments that can be passed to tools.
+ToolArguments = Dict[str, Any]
+T = TypeVar("T", bound="Tool")
+
+
+class ToolResult(TypedDict):
+    """
+    The base result dictionary of any tool.
+    """
+    result: Result
+
+
+class PingResult(ToolResult):
+    """
+    The result dictionary of the 'ping' tool.
+    """
+    #: Host(s) that are alive
```

Comment:
For the results of `ping`:
* `min_time` and `max_time` are probably the most useful and also the easiest to use with some `Comparison` (`<= 500ms` for example)
* Values that return a list (as in `alive` or `unresponsive`) or a dictionary (as in `times`) are probably not easy to use with `Comparison`
   * Does that indicate that the tool definition is poor?
   * Does it indicate that new comparisons may be required?

I haven't really decided yet, but I'm throwing this out there

Reply:
I mitigated this sub-keys being valid in the identifier list (e.g., `hosts.hostname` could be valid).
The splitting method can still use some work as we wouldn't want `hosts.google.com` to turn into `("hosts", "google", "com")`.

Reply:
It may be useful to summarize difficult to compare keys e.g. `num_alive`, `num_unresponsive`. If we want to support comparisons of sequence types, then maybe we could also include reduction options like `len` that drop them down to a scalar more generically?

Reply:
Added `num_*` variants. I think we can probably push back the decision on the reduction options for other tools.

A _better_ simple check here could just be `num_unresponsive == 0`